perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/ConsolidatorManagerImpl.java

```diff
@@ -460,6 +458,184 @@ public List<UserExtSource> consolidateIdentityUsingToken(PerunSession sess, Stri
+				StringUtils.join(Arrays.asList(userInfo2.sub, userInfo2.issuer, sess.getPerunPrincipal().getExtSourceType()), " | "));
+
+		}
+
+		AuthzResolverBlImpl.refreshSession(sess);
+
+		if (BeansUtils.getCoreConfig().isSendIdentityAlerts()) {
+			try {
+				Utils.sendIdentityAddedAlerts(sess, createdUes);
+			} catch (Exception e) {
+				log.error("Failed to send identity added alerts.", e);
+			}
+		}
+		return null;
+	}
+
+	private record OidcUserInfo(String sub, String issuer, Object userInfo) {}
+
+	private static OidcUserInfo getOidcSubIssuer(String accessToken, String issuer) throws PerunException {
+		RestTemplate restTemplate = new RestTemplate();
+		JsonNode config = restTemplate.getForObject(issuer + "/.well-known/openid-configuration", JsonNode.class);
+		HttpHeaders headers = new HttpHeaders();
+		headers.setBearerAuth(accessToken);
+		HttpEntity<Void> entity = new HttpEntity<>(headers);
+
+		ResponseEntity<JsonNode> userInfoResponse;
+		try {
+			userInfoResponse = restTemplate.exchange(config.path("userinfo_endpoint").textValue(), HttpMethod.GET, entity, JsonNode.class);
+		} catch (HttpClientErrorException ex) {
+			if(ex.getStatusCode() ==  HttpStatus.FORBIDDEN) {
+				log.info("Token {} for joining identities is no longer valid for issuer: {}", accessToken, issuer);
+				InvalidTokenException iex = new InvalidTokenException("Your token for joining identities is no longer valid. Please retry from the start.");
+				throw iex;
+			} else {
+				log.info("Failed to get userInfoResponse for access token: {}. The response code was: {}", accessToken, ex.getStatusCode());
+				InternalErrorException iex = new InternalErrorException("Failed to get userInfoResponse for access token: " + accessToken + ". The response code was: " + ex.getStatusCode());
+				throw iex;
+			}
+
+		}
+		var userInfo = userInfoResponse.getBody();
+		var originalIssuer = userInfo.path("target_issuer").asText();
+		if(isEmpty(originalIssuer)) {
+			log.error("issuer was empty or null: {}", originalIssuer);
+			throw new InternalErrorException("sub and issuer were null: " + originalIssuer);
+		}
+
+		var login = userInfo.path("eduperson_unique_id").asText();
+		if(isEmpty(login)) {
+			login = userInfo.path("eduperson_principal_name").asText();
+			if(isEmpty(login)) {
+				login = userInfo.path("saml2_nameid_persistent").asText();
+				if(isEmpty(login)) {
+					var edupersonTargetedId = userInfo.path("eduperson_targeted_id").get(0);
+					login = (edupersonTargetedId == null) ? "" : edupersonTargetedId.asText();
+					if(isEmpty(login)) {
+						var voperson_external_id = userInfo.path("voperson_external_id").get(0);
+						login = (voperson_external_id == null) ? "" : voperson_external_id.asText();
+					}
+				}
+			}
+		}
+		return new OidcUserInfo(login, originalIssuer, userInfo);
+	}
+
+	private static void getAdditionalInformations(String accessToken, Map<String, String> additionalInformations) {
```

Comment:
the same code as in the Api class - needs to be refactored so we don't have completely the same code on the two places